src/solidity.ts

```diff
@@ -191,6 +205,27 @@ abstract class SolidityContractItem implements Linkable {
   }
 }
 
+class SolidityVariable extends SolidityContractItem {
```

Comment:
@frangio:

First of all, "doesn't have the function arguments" isn't entirely true.
Consider the following example:
```
contract Contract1 {
    uint[42]                          public x1; // function x1(uint)
    mapping (address => uint)         public x2; // function x2(address)
    mapping (address => uint[42])     public x3; // function x3(address, uint)
    mapping (address => uint)[42]     public x4; // function x4(uint, address)
    mapping (address => uint[42])[84] public x5; // function x5(uint, address, uint)
}

contract Contract2 {
    uint public x1;
    uint public x2;
    uint public x3;
    uint public x4;
    uint public x5;

    constructor() public {
        uint u = 1;
        address a = address(1);
        Contract1 contract1 = new Contract1();
        x1 = contract1.x1(u);
        x2 = contract1.x2(a);
        x3 = contract1.x3(a, u);
        x4 = contract1.x4(u, a);
        x5 = contract1.x5(u, a, u);
    }
}
```
I initially had an idea, which was indeed based on the notion that compiler-generated getter functions take no arguments:
- Your tool will provide just the ability to list the variables (which is pretty much what I had already completed in this PR)
- In the input template file, one could simply do something like:
```
{{#ownVariables}}
# Variable `{{name}}() → {{type}}`
{{/ownVariables}}
```
But then I realized that it wouldn't do the job right for all types of variables (in particularly, arrays and mappings).

Second, I don't quite understand the gain in not extending `SolidityContractItem`, since functions `name`, `fullName` and `anchor` (which relies on `signature` which we override) should actually be implemented exactly the same.
It is true that the `args` function requires AST information which is not available for variables, but we can simply override it in the same way that we override the `natspec` function (though the latter will (hopefully) be overridden temporarily, while the former will be overridden permanently).
This also addresses the suggestion that you've made about the `signature` function.
I believe that this function is actually very much relevant for variables.
First, because a variable-declaration can be considered a signature just as much as a function-declaration can.
Second, because on the technical aspect, it is used by the `anchor` function, which I think we should support for variables.

Please let me know what you think.

Thanks

Reply:
>First of all, "doesn't have the function arguments" isn't entirely true.

What I mean is that the AST doesn't have the information about the function arguments.

I completely agree that it is relevant information. But since the AST doesn't make it available, we would have to generate it ourselves, reproducing the same logic that the Solidity compiler uses. I would prefer not to do that right now, and for the moment only expose the list of variables and their type and visibility.

Note that this PR as it is now doesn't implement the logic for figuring out what are the arguments for a generated getter. There's some tricky logic there, so let's leave that for later on, and for now focus on make state variable information available.

Reply:
> Second, because on the technical aspect, it is used by the anchor function, which I think we should support for variables.

We should support the `anchor` function, that's why I suggested to implement the `Linkable` interface directly, without relying on `SolidityContractItem` that requires all of the getter information we don't yet have.

Reply:
@frangio:
OK, done, but I'm a little confused about whether or not we want to expose function `signature`...?
And if not, then (for the exact same reason) why should we expose function `natspec`?
Thanks

Reply:
If you don't mind I will commit on top of this branch to show what I mean regarding `signature` and `anchor`.

Reply:
@frangio:
Sure. Please let me know if anything is required on my side.
Thanks.